### PR TITLE
Theano versions of MNIST tutorial and FGSM

### PR DESCRIPTION
**Pull request by mahnerak**
1. Separate attacks modules for Tensor and Theano
2. Theano implementation of FGSM
3. New tutorial for FGSM in Theano

Next steps would be:
- Remove Keras dependency in JSMA implementation
- Move present JSMA implementation to attacks_tf
- Implement JSMA in Theano
- Implement more attacks

### CODE REVIEW

File: cleverhans/attacks.py [hunk 1]
```diff
@@ -35,48 +35,12 @@ def fgsm(x, predictions, eps, back='tf', clip_min=None, clip_max=None):
...
-
-    # If clipping is needed, reset all values outside of [clip_min, clip_max]
-    if (clip_min is not None) and (clip_max is not None):
-        adv_x = tf.clip_by_value(adv_x, clip_min, clip_max)
-
-    return adv_x
-
+        # Compute FGSM using Theano
+        from .attack_th import fgsm as fgsm_th
+        raise fgsm_th(x, predictions, eps, clip_min=clip_min, clip_max=clip_max)
```
Comment (goodfeli) [inline cleverhans/attacks.py]: This should be `return` rather than `raise`. Would you mind adding a small smoke test that would have failed for this?

File: cleverhans/attacks_tf.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    scaled_signed_grad = eps * signed_grad
+
+    # Add perturbation to original example to obtain adversarial example
+    adv_x = tf.stop_gradient(x + scaled_signed_grad)
+
+    # If clipping is needed, reset all values outside of [clip_min, clip_max]
+    if (clip_min is not None) and (clip_max is not None):
+        adv_x = tf.clip_by_value(adv_x, clip_min, clip_max)
+
+    return adv_x
```
Comment (goodfeli) [inline cleverhans/attacks_tf.py]: github isn't happy with the whitespace here

File: cleverhans/utils_th.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                  T.sqrt(accu_new + epsilon))
+        updates[param] = param - learning_rate * update
+
+        # update delta_accu (as accu, but accumulating updates)
+        delta_accu_new = rho * delta_accu + (one - rho) * update ** 2
+        updates[delta_accu] = delta_accu_new
+
+    return updates
+
+def th_model_loss(y, model, mean=True):
```
Comment (goodfeli) [inline cleverhans/utils_th.py]: shouldn't need the `th_` suffix if we're already inside a `th_` model

**Comment by goodfeli** [comment]
Could you make a `test_mnist_accuracy_th` or similar? This looks like a lot of new code without new tests to accompany it. Otherwise looks like basically the right direction, thanks!

**Comment by mahnerak** [comment]
i'm working on this right now
theano and TF results are different now, probably because of different initializations

File: cleverhans/utils_th.py (hunk 3 above)
Comment (mahnerak) [inline cleverhans/utils_th.py]: The `th_` suffix removed

File: cleverhans/attacks_tf.py (hunk 2 above)
Comment (mahnerak) [inline cleverhans/attacks_tf.py]: Fixed!

**Comment by mahnerak** [comment]
`tests/test_mnist_accuracy_th.py` is provided. It tests not only the accuracy of trained MNIST model, but also accuracy of the same model on the crafted adversarial examples.

**Comment by npapernot** [comment]
I will commit a few changes to the documentation and make some minor changes to the code. Should be done today and will merge this tonight. Thanks @mahnerak for the contribution, this is a good first step towards supporting TF and TH.

**Comment by npapernot** [comment]
a6745b0 moves JSMA to attacks_tf

**Comment by npapernot** [comment]
LGTM
